### PR TITLE
[Issue #79] RestClientBuilderListener for global provider registration

### PR DESCRIPTION
**Pull request by andymc12**
This would resolve issue #79.

Signed-off-by: Andy McCright <j.andrew.mccright@gmail.com>

### CODE REVIEW

File: api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/*
+ *******************************************************************************
+ * Copyright (c) 2017-2018 Contributors to the Eclipse Foundation
```
[the remark this answers is not in the document]
Comment (andymc12) [reply api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java]: Will fix - good catch.

**Comment by andymc12** [comment]
@Emily-Jiang I chose the ServiceLoader pattern so that user apps could use this approach too. The use case for that would be similar to OpenTracing: suppose a user wants to log all outbound MP Rest Client calls - but maybe not using MP OpenTracing.  They could do their own logging in a filter in a similar fashion.  There are other use cases - like adding a custom header to all requests coming from a particular client endpoint, or registering a default set of ResponseExceptionMappers that are common across an entire application (thus avoiding the more verbose option of explicitly registering each mapper with each interface).

Could you send me a link to more info on the MP listener/registration pattern?  If it would work for users too, then it might be preferable to the ServiceLoader pattern that I have implemented here.

**Comment by johnament** [comment]
There is no consistent listener/registration pattern.  I know something was just added recently to ConfigJSR when you add a listener, but none of the actual MP specs use it.

File: api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * </code> file in an archive on the current thread's context classloader's
+ * class path.
+ *
+ * Note that the <code>newBuilder</code> method will be called when the
+ * RestClientBuilder is constructed, not when it's <code>build</code> method is
+ * invoked.  This allows the caller to override global providers if they desire.
+ */
+public interface ProviderRegistrar {
+
+    void newBuilder(RestClientBuilder builder);
```
Comment (johnament) [inline api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java]: Though I do see what Emily means.  the method `newBuilder` is a life cycle method when a new builder is created.   There's a few other things we could do:

- Create an `AutoRegisteredClientFeature` interface which extends `Feature`, used on bootstrap to load features that should always be applied.
- Rename this method from `newBuilder` to `onNewBuilder` or even `customize` to better explain its purpose.

File: spec/src/main/asciidoc/providers.asciidoc [hunk 3]
```diff
@@ -90,6 +90,8 @@ Any methods that read the response body as a stream must ensure that they reset
 
 In addition to defining providers via the client definition, interfaces may use the `@RegisterProvider` annotation to define classes to be registered as providers in addition to providers registered via the `RestClientBuilder`
 
+Providers may also be registered by implementing the `ProviderRegistrar` interface.  This interface is intended as an SPI to allow global provider registration.  The implementation of this interface must be specified in a META-INF/services/org.eclipse.microprofile.rest.client.spi.ProviderRegistrar file, following the service loader pattern.
```
Comment (johnament) [inline spec/src/main/asciidoc/providers.asciidoc]: The `META-INF` file should be listed as a code.

File: api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java (hunk 2 above)
Comment (andymc12) [reply api/src/main/java/org/eclipse/microprofile/rest/client/spi/ProviderRegistrar.java]: Makes sense.  Personally, I like `onNewBuilder` the best.  Will change.

File: api/src/main/java/org/eclipse/microprofile/rest/client/spi/RestClientBuilderListener.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.rest.client.spi;
+
+import org.eclipse.microprofile.rest.client.RestClientBuilder;
+
+/**
+ * Implementations of this interface will be notified when new RestClientBuilder
+ * instances are being contructed.  This will allow implementations to register
```
Comment (Emily-Jiang) [inline api/src/main/java/org/eclipse/microprofile/rest/client/spi/RestClientBuilderListener.java]: constructed

**Comment by Emily-Jiang** [comment]
@johnament was right. We discussed this in configjsr. @struberg recently ported this back to MP Config. https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/Config.java I understand a bit more of your use case. Maybe service loader pattern fits more as it is automatic.

**Comment by andymc12** [comment]
Thanks @Emily-Jiang - I updated the version to 1.1 in the package-info.java file and corrected the typo.  Thanks for catching these issues.

**Comment by andymc12** [comment]
Merged with other changes - and will merge to master per John's review in gitter channel.

**Pull request merged**